### ISSUE TITLE
Text Detection training code(craft) has BUG

### ISSUE THREAD

**Issue by masoudMZB** [JaidedAI/EasyOCR#1005]
I've tried to train craft with the Craft training code provided by the easyOCR repository, but I think the code has a bug. I trained with one sample image of the ICDAR dataset ( img_990.jpg, my training and validation dataset for finetuning was this image, just one image ),  I used the pre-trained model of craft for finetuning, after 5 epoch model will learn to predict nothing ( list of prediction boxes become empty) but training loss will reduce. How is that possible? I logged more info inside the code and I print them below : 

> results :  [{'precision': 0.5, 'recall': 1.0, 'hmean': 0.6666666666666666, 'pairs': [{'gt': 0, 'det': 1}], 'iouMat': [[0.11546084772668544, 0.5038291537164455]], 'gtPolPoints': [array([[  9,  55],
       [168,  55],
       [168,  19],
       [  9,  19]], dtype=int32)], 'detPolPoints': [array([[10.285714, 24.      ],
       [35.42857 , 24.      ],
       [35.42857 , 50.285713],
       [10.285714, 50.285713]], dtype=float32), array([[ 56.      ,  24.      ],
       [165.71428 ,  24.      ],
       [165.71428 ,  50.285713],
           [ 56.      ,  50.285713]], dtype=float32)], 'gtCare': 1, 'detCare': 2, 'gtDontCare': [], 'detDontCare': [], 'detMatched': 1, 'evaluationLog': 'GT polygons: 1\nDET polygons: 2\nMatch GT #0 with Det #1\n'}]
    ----------------------------------------
    ----------------------------------------
    {'precision': 0.5, 'recall': 1.0, 'hmean': 0.6666666666666666, 'numGlobalCareDet': 2, 'numGlobalCareGt': 1, 'matchedSum': 1}
    ./data_root_dir/ch4_training_images/task_data_0.jpg
    /usr/local/lib/python3.10/dist-packages/torch/nn/_reduction.py:42: UserWarning: size_average and reduce args will be deprecated, please use reduction='none' instead.
      warnings.warn(warning.format(ret))
    this is what is sending to batch_image_loss 
    loss_region:tensor([[[0.0000e+00, 0.0000e+00, 0.0000e+00,  ..., 1.8176e-04,
              1.8178e-04, 1.8013e-04],
             [0.0000e+00, 0.0000e+00, 0.0000e+00,  ..., 1.8177e-04,
              1.8180e-04, 1.8281e-04],
             [0.0000e+00, 0.0000e+00, 0.0000e+00,  ..., 1.8182e-04,
              1.8186e-04, 1.9605e-04],
             ...,
             [1.1542e-01, 1.1531e-01, 1.1127e-01,  ..., 9.0923e-05,
              9.0928e-05, 9.1294e-05],
             [1.0864e-01, 1.1066e-01, 1.0959e-01,  ..., 9.0938e-05,
              9.1470e-05, 9.3088e-05],
             [1.1095e-01, 1.0960e-01, 1.0875e-01,  ..., 8.0241e-05,
              8.8683e-05, 1.0420e-04]]], device='cuda:0', grad_fn=<MulBackward0>)
    region_scores_label:tensor([[[0.0000, 0.0000, 0.0000,  ..., 0.0000, 0.0000, 0.0000],
             [0.0000, 0.0000, 0.0000,  ..., 0.0000, 0.0000, 0.0000],
             [0.0000, 0.0000, 0.0000,  ..., 0.0000, 0.0000, 0.0000],
             ...,
             [0.4985, 0.4937, 0.4852,  ..., 0.0000, 0.0000, 0.0000],
             [0.4948, 0.4901, 0.4816,  ..., 0.0000, 0.0000, 0.0000],
             [0.4926, 0.4878, 0.4794,  ..., 0.0000, 0.0000, 0.0000]]],
           device='cuda:0')
    neg_rto:0.3
    n_min_neg:5000
    Let's see what 
    here checking what is positive pixel number : 18774.0
    here checking what is negative_pixel_number : 128682.0
    positive loss is 0.10362127423286438
    negative loss is 0.0016526866238564253
    Let's see what 
    here checking what is positive pixel number : 26033.0
    here checking what is negative_pixel_number : 121423.0
    positive loss is 0.1118001639842987
    negative loss is 0.0015156574081629515
    lets see what is char_loss 0.10527396202087402 and affi_loss 0.11331582069396973
    2023-05-01:14:19:05, training_step: 5|15, learning rate: 0.00010000, training_loss: 0.21859, avg_batch_time: 17.51073
    Saving state, index: 5
    100% 1/1 [00:00<00:00, 56.85it/s]
    ------------------------------------------------------------
    ------results :  [{'precision': 0.0, 'recall': 0.0, 'hmean': 0, 'pairs': [], 'iouMat': [[0.4718834949057586]], 'gtPolPoints': [array([[  9,  55],
           [168,  55],
           [168,  19],
           [  9,  19]], dtype=int32)], 'detPolPoints': [array([[ 57.142857,  24.      ],
           [164.57143 ,  24.      ],
           [164.57143 ,  49.142857],
           [ 57.142857,  49.142857]], dtype=float32)], 'gtCare': 1, 'detCare': 1, 'gtDontCare': [], 'detDontCare': [], 'detMatched': 0, 'evaluationLog': 'GT polygons: 1\nDET polygons: 1\n'}]
    ----------------------------------------
    ----------------------------------------
    {'precision': 0.0, 'recall': 0.0, 'hmean': 0, 'numGlobalCareDet': 1, 'numGlobalCareGt': 1, 'matchedSum': 0}
    ./data_root_dir/ch4_training_images/task_data_0.jpg
    /usr/local/lib/python3.10/dist-packages/torch/nn/_reduction.py:42: UserWarning: size_average and reduce args will be deprecated, please use reduction='none' instead.
      warnings.warn(warning.format(ret))
    this is what is sending to batch_image_loss 
    loss_region:tensor([[[2.3100e-07, 4.2590e-06, 7.5753e-06,  ..., 9.4357e-05,
              9.4357e-05, 9.8660e-05],
             [4.6670e-06, 6.8195e-06, 1.0874e-05,  ..., 9.4345e-05,
              9.4349e-05, 9.6100e-05],
             [8.9083e-06, 1.1795e-05, 1.6938e-05,  ..., 9.4362e-05,
              9.4369e-05, 9.5385e-05],
             ...,
             [1.7950e-04, 1.8878e-04, 1.8876e-04,  ..., 0.0000e+00,
              0.0000e+00, 0.0000e+00],
             [1.9854e-04, 1.8880e-04, 1.8878e-04,  ..., 0.0000e+00,
              0.0000e+00, 0.0000e+00],
             [2.0348e-04, 2.3584e-04, 2.0312e-04,  ..., 0.0000e+00,
              0.0000e+00, 0.0000e+00]]], device='cuda:0', grad_fn=<MulBackward0>)
    region_scores_label:tensor([[[0.0160, 0.0167, 0.0176,  ..., 0.0000, 0.0000, 0.0000],
             [0.0168, 0.0174, 0.0184,  ..., 0.0000, 0.0000, 0.0000],
             [0.0180, 0.0186, 0.0196,  ..., 0.0000, 0.0000, 0.0000],
             ...,
             [0.0000, 0.0000, 0.0000,  ..., 0.0000, 0.0000, 0.0000],
             [0.0000, 0.0000, 0.0000,  ..., 0.0000, 0.0000, 0.0000],
             [0.0000, 0.0000, 0.0000,  ..., 0.0000, 0.0000, 0.0000]]],
           device='cuda:0')
    neg_rto:0.3
    n_min_neg:5000
    Let's see what 
    here checking what is positive pixel number : 20353.0
    here checking what is negative_pixel_number : 127103.0
    positive loss is 0.06648588925600052
    negative loss is 0.0022214415948837996
    Let's see what 
    here checking what is positive pixel number : 41201.0

    here checking what is negative_pixel_number : 106255.0
    positive loss is 0.10165758430957794
    negative loss is 0.0013221692061051726
    lets see what is char_loss 0.06870733201503754 and affi_loss 0.10297975689172745
    2023-05-01:14:19:07, training_step: 6|15, learning rate: 0.00010000, training_loss: 0.17169, avg_batch_time: 18.67954
    Saving state, index: 6
    100% 1/1 [00:00<00:00, 27.06it/s]

This is not correct. I don't know if the problem is inside implemented Loss function or any other part of the code.

I'll attach image and corresponding gt here 

[gt_img_990.txt](https://github.com/JaidedAI/EasyOCR/files/11379963/gt_img_990.txt)
![img_990](https://user-images.githubusercontent.com/33782977/235846111-c4c88fcf-9138-411e-8c37-cd3179077447.jpg)

**Comment by MdotO** [comment]
From my observations and understanding, having a lower train loss but with no boxes/worse precision /recall is quite understandable as the model is still training + I see from the logs, that model first output **two separate detection boxes** instead of 1 initially. Afterwards, it produced one bounding box as expected in iteration 5 but because of < 0.5 IOU , it was not used and hence precision and recall = 0. After training result is very similar to og model result in that it atleast learns to remove one of original detection box and use the second with slightly worse IOU. Overall, this lead to an improvement in train loss but because of IOU threshold, didnt lead to a positive detection

Some suggestions/confirmations from my side for the above :
1) Just confirming whether you ran the training code multiple times for reproducibility ?
2) Did you try seeing after e.g 20 epochs(or iterations in ur case as there is just 1 image)  ?
3) You can try drawing out the bbox for the og weights and trained weights you get and set NMS/ bbox threshold to be lower than 0.5. Based on the results, it should show 1 bbox though not fully aligned with ur groundthruth(but it's getting there). The eval script contains the code for drawing I believe - just need to reduce threshold

**Comment by kulsoom-abdullah** [comment]
Hi @masoudMZB,

I noticed your issue with training Craft using the provided code and the unexpected results you encountered. I faced a similar issue a few months ago when attempting to use a fine-tuned CRAFT model with EasyOCR. To summarize:

- By default, EasyOCR uses and downloads the [default model weights](https://github.com/JaidedAI/EasyOCR/issues/825), even if a custom one is provided.
- Adjusting thresholds, such as text_threshold and low_text,  in the [YAML config](trainer/craft/config/custom_data_train.yaml)didn't make a difference.
- The ground truth format in the README led us to think about how bounding boxes for individual characters versus whole phrases might make a difference.
- Based on your logs, I see certain inconsistencies as well, particularly in the training loss and prediction results.

I'm wondering if others have faced this and found a solution, or if there's insight into potential bugs in the training code or the loss function. Any guidance would be greatly appreciated!

**Comment by wesleylp** [comment]
hey everyone!
I'm facing the same issue. Does anyone got a solution?

**Comment by JamesScott4520** [comment]
I'm facing the same issue.
2024-10-28:11:46:25, training_step: 4400|30000, learning rate: 0.00010000, training_loss: 0.04127, avg_batch_time: 2012.95587
Saving state, index: 4400
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00, 30.81it/s]
{'precision': 0, 'recall': 0.0, 'hmean': 0}

I have no idea what is going on。
already trained 4400 epoch still {'precision': 0, 'recall': 0.0, 'hmean': 0}
just trained one image……

**Comment by AliesTaha** [comment]
@JamesScott4520 I have the exact same problem you are facing. Been wrestling with it for the last week. One image, training works well. Minimal loss, yet 0 accuracy. Any updates on how you faced it?
